design/0001-frr.md

```diff
@@ -0,0 +1,340 @@
+and [here](https://metallb.universe.tf/configuration/kube-router/) for the
+Kube-Router case. This enhancement may be a step towards resolving these
+issues by potentially allowing MetalLB to reuse a BGP agent already in use on
+the node. This local agent could then establish the session with the remote
+agent and be the single entity peering with the remote agent. In this way,
+MetalLB and other software on the node could share the BGP session with the
+remote agent. For the Calico case, this would require integrating BIRD as an
+alternative BGP implementation as Calico uses BIRD for its BGP functionality.
+This would also require additional work to configure FRR in a way that would
+allow it to be shared.
+
+### Goals
+
+* Provide a configuration option to allow use of alternate BGP agents in a
+  MetalLB deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation in order to enable
+  the eventual retirement of the current native BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+* Add additional documentation to indicate how to troubleshoot and debug any
+  issues with FRR.
+* This will not expose direct configuration of FRR to the MetalLB user.
+
+### Non-Goals
+
+* Coexistence of two types of BGP agents (e.g. MetalLB native BGP agent AND FRR
+  BGP agent) is not a goal of this work.
+* The purpose of this is not to provide a method to integrate arbitrary BGP
+  stacks into MetalLB. However, this work will develop a well-defined
+  interface that could be used to integrate other BGP stacks and will be a
+  template to do so.
+* This will not support runtime changing of BGP agent type.
+* This will not extend the functionality of the MetalLB BGP implementation in
+  in order to take advantage of the additional cababilities of FRR. This will
+  be a goal of future enhancements.
+* This will not explicitly support allowing another actor to interact with FRR.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer, I want to be able to run a comprehensive set of BGP system
+tests, preferably based on kubernetes e2e tests, in order to ensure the MetalLB
+BGP implementation works as expected.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to specify a configuration option
+when deploying MetalLB in order to select the MetalLB BGP agent to be used by
+MetalLB
+
+#### Story 3
+
+As a developer, I want to be able to configure FRR from MetalLB using FRR Go
+bindings in order to improve performance, reliability and debugging
+capabilities.
+
+### Risks and Mitigations
+
+FRR does not provide documented northbound Go bindings to allow configuration of
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
+mitigate this risk, we can wrap the FRR ‘vtysh’ command line interface in Go
```

Comment:
I'm not sure. I do agree with you that gut-feeling would be that it is stable as it would need to be scriptable. However, I would concede that this is a (low) risk item.

Reply:
Cool, thanks!